src/Microsoft.AspNetCore.Routing.Abstractions/MetadataCollection.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class EndpointMetadataCollection : IReadOnlyList<object>
+    {
+        public static readonly EndpointMetadataCollection Empty = new EndpointMetadataCollection(Array.Empty<object>());
+
+        private readonly object[] _items;
+
+        public EndpointMetadataCollection(IEnumerable<object> items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _items = items.ToArray();
+        }
+
+        public object this[int index] => _items[index];
+
+        public int Count => _items.Length;
+
+        public T GetMetadata<T>() where T : class
+        {
+            for (var i = _items.Length -1; i >= 0; i--)
+            {
+                var item = _items[i] as T;
+                if (item !=null)
+                {
+                    return item;
+                }
+            }
+
+            return default;
+        }
+
+        public IEnumerable<T> GetOrderedMetadata<T>() where T : class
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    yield return item;
```

Comment:
I also expect this code to change a bit. The main usage of this boy is for the kinds of things filters do in MVC. You sometimes want to get all policies of a type, and sometimes want to get the 'best' policy.

I think we'll eventually make this type more dictionary-like, which is the reason for the nominal type rather than IReadOnlyList<T>